### PR TITLE
feat: add CJK friendly emphasis extension

### PR DESCRIPTION
**Pull request by 3w36zj6**
CommonMark has a problem that the following emphasis marks `**` are not recognized as emphasis marks in CJK.

```md
**このアスタリスクは強調記号として認識されず、そのまま表示されます。**この文のせいで。

**该星号不会被识别，而是直接显示。**这是因为它没有被识别为强调符号。

**이 별표는 강조 표시로 인식되지 않고 그대로 표시됩니다(이 괄호 때문에)**이 문장 때문에.
```

This pull request introduces support for CJK-friendly emphasis handling in the Markdown parser, aligning with the [CommonMark CJK-friendly amendments specification](https://github.com/tats-u/markdown-cjk-friendly/blob/main/specification.md). 

It adds a new option to enable CJK-friendly emphasis parsing, updates the delimiter run logic to properly handle CJK characters and punctuation, and includes comprehensive tests to verify the new behavior. By default, the feature is disabled to maintain backward compatibility.

In addition to the specification, I also refer to the [Tips for Implementers](https://github.com/tats-u/markdown-cjk-friendly/blob/main/implementers-tips.md) and [Concrete ranges of each terms](https://github.com/tats-u/markdown-cjk-friendly/blob/main/ranges.md) in [tats-u/markdown-cjk-friendly](https://github.com/tats-u/markdown-cjk-friendly) for implementation.

### CODE REVIEW

File: pulldown-cmark/src/firstpass.rs [hunk 1]
```diff
@@ -2248,6 +2252,36 @@ fn fixup_end_of_definition_list(tree: &mut Tree<Item>, list_ix: TreeIndex) {
...
+fn previous_two_chars(s: &str, ix: usize) -> (Option<char>, Option<char>) {
+    let mut iter = s[..ix].chars();
+    let mut prev_prev = None;
+    let mut prev = None;
+    while let Some(ch) = iter.next() {
+        prev_prev = prev;
+        prev = Some(ch);
+    }
+    (prev, prev_prev)
+}
```
[the remark this answers is not in the document]
Comment (ollpu) [reply pulldown-cmark/src/firstpass.rs]: Related, maybe this should not take `ix`, so that it's the caller's responsibility to slice the string.

File: pulldown-cmark/tests/html.rs [hunk 2]
```diff
@@ -316,6 +316,88 @@ fn trim_space_nl_at_end_of_paragraph() {
...
+    let mut output = String::new();
+    html::push_html(&mut output, Parser::new_ext(input, opts));
+
+    assert_eq!(expected, output);
+}
+
+#[test]
+fn cjk_friendly_emphasis_simplified_chinese_punctuation() {
+    let input = "这是。**强调**吗？";
+    let expected = "<p>这是。<strong>强调</strong>吗？</p>\n";
```
Comment (tats-u) [inline pulldown-cmark/tests/html.rs]: This is not specific to this extension.

File: pulldown-cmark/tests/html.rs [hunk 3]
```diff
@@ -316,6 +316,88 @@ fn trim_space_nl_at_end_of_paragraph() {
...
+    let mut output = String::new();
+    html::push_html(&mut output, Parser::new_ext(input, opts));
+
+    assert_eq!(expected, output);
+}
+
+#[test]
+fn cjk_friendly_emphasis_handles_cjk_suffix() {
+    let input = "_強調_箇所";
+    let expected = "<p><em>強調</em>箇所</p>\n";
```
Comment (tats-u) [inline pulldown-cmark/tests/html.rs]: This cannot be fixed even by this extension.

File: pulldown-cmark/tests/html.rs [hunk 4]
```diff
@@ -316,6 +316,88 @@ fn trim_space_nl_at_end_of_paragraph() {
...
+    let mut output = String::new();
+    html::push_html(&mut output, Parser::new(input));
+
+    assert_eq!(expected, output);
+}
+
+#[test]
+fn cjk_friendly_emphasis_enabled() {
+    let input = "これは_強調_です";
+    let expected = "<p>これは<em>強調</em>です</p>\n";
```
Comment (tats-u) [inline pulldown-cmark/tests/html.rs]: This cannot be fixed even by this extension.

File: pulldown-cmark/src/firstpass.rs [hunk 5]
```diff
@@ -2398,23 +2424,44 @@ fn delim_run_can_open(
...
     // `*`, `~~`, and `^` can be intraword, `~` can only be interword if it's subscript, `_` cannot
-    if (delim == b'*' || delim == b'^') && !is_punctuation(next_char) {
+    if (delim == b'*' || delim == b'^') && !restricted_next {
         return true;
     }
     if delim == b'~' && run_len > 1 {
         return true;
     }
-    let prev_char = s[..ix].chars().last().unwrap();
+    let (prev_char_opt, prev_prev_char) = previous_two_chars(&s[..ix]);
```
Comment (tats-u) [inline pulldown-cmark/src/firstpass.rs]: Don't have pulldown-cmark retrieve `prev_prev_char` unless it's actually needed in this extension (especially when it has not been turned on)

File: pulldown-cmark/src/cjk.rs [hunk 6]
```diff
@@ -0,0 +1,87 @@
+use crate::puncttable::is_punctuation;
+
+/// Returns true when `ch` is considered a CJK character by the
+/// CommonMark CJK-friendly amendments specification.
+pub(crate) fn is_cjk_character(ch: char) -> bool {
```
Comment (tats-u) [inline pulldown-cmark/src/cjk.rs:8]: Add Unicode version. If it is lower than 17, update the code to 17.

File: pulldown-cmark/src/firstpass.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (tats-u) [reply pulldown-cmark/src/firstpass.rs]: .char() returns DoubleEndedIterator.

https://doc.rust-lang.org/src/core/str/iter.rs.html#125
https://stdrs.dev/nightly/x86_64-pc-windows-gnu/src/core/str/validations.rs.html#79-113

File: pulldown-cmark/tests/html.rs [hunk 7]
```diff
@@ -316,6 +316,88 @@ fn trim_space_nl_at_end_of_paragraph() {
...
+
+    let mut opts = Options::empty();
+    opts.insert(Options::ENABLE_CJK_FRIENDLY_EMPHASIS);
+
+    let mut output = String::new();
+    html::push_html(&mut output, Parser::new_ext(input, opts));
+
+    assert_eq!(expected, output);
+}
+
```
Comment (tats-u) [inline pulldown-cmark/tests/html.rs]: The current test cases are currently redundant and far from sufficient. You need to add edge cases (e.g. ones using variation selectors or decomposed Hangul parts) and cases where an `*`/`**` is placed between an alphabet and a CJK punctuation character.

File: pulldown-cmark/src/lib.rs [hunk 8]
```diff
@@ -768,6 +769,12 @@ bitflags::bitflags! {
         const ENABLE_WIKILINKS = 1 << 15;
         /// Colon-delimited Container Extension Blocks.
         const ENABLE_CONTAINER_EXTENSIONS = 1 << 16;
+        /// CommonMark CJK-friendly emphasis heuristics.
+        ///
+        /// Aligns delimiter run handling with the
+        /// [CommonMark CJK-friendly amendments](https://github.com/tats-u/markdown-cjk-friendly/blob/main/specification.md),
+        /// enabling emphasis between CJK characters without surrounding whitespace.
```
Comment (tats-u) [inline pulldown-cmark/src/lib.rs]:
```suggestion
        /// [CommonMark CJK-friendly amendments/extension](https://github.com/tats-u/markdown-cjk-friendly/blob/main/specification.md),
        /// enabling emphasis around CJK characters without surrounding whitespace.
```

File: pulldown-cmark/src/firstpass.rs [hunk 9]
```diff
@@ -2472,18 +2523,32 @@ fn delim_run_can_close(
...
-    next_char.is_whitespace() || is_punctuation(next_char)
+    if !cjk_friendly {
+        next_char.is_whitespace() || is_punctuation(next_char)
+    } else if prev_non_cjk_seq {
+        next_char.is_whitespace()
+            || is_non_cjk_punctuation_character(next_char)
+            || is_cjk_character(next_char)
+    } else {
+        true
+    }
```
Comment (tats-u) [inline pulldown-cmark/src/firstpass.rs]:
```suggestion
    next_char.is_whitespace() || if !cjk_friendly {
        is_punctuation(next_char)
    } else {
        !prev_non_cjk_seq
            || is_non_cjk_punctuation_character(next_char)
            || is_cjk_character(next_char)
    }
```

File: pulldown-cmark/src/firstpass.rs [hunk 10]
```diff
@@ -2428,16 +2460,31 @@ fn delim_run_can_open(
...
+    };
+    let prev_cjk_seq = cjk_friendly && is_prev_cjk_sequence(prev_char_opt, prev_prev_char);
+    let prev_is_ivs = cjk_friendly && is_ideographic_variation_selector(prev_char);
+    let prev_base_char = base_char_for_sequence(prev_char_opt, prev_prev_char);
+    let blocked_by_quote = matches!(prev_base_char, Some(']') | Some(')'));
+
     prev_char.is_whitespace()
-        || is_punctuation(prev_char) && (delim != b'\'' || ![']', ')'].contains(&prev_char))
+        || (prev_non_cjk_seq && (delim != b'\'' || !blocked_by_quote))
+        || (cjk_friendly && (prev_cjk_seq || prev_is_ivs))
```
Comment (tats-u) [inline pulldown-cmark/src/firstpass.rs]: Utilize `if` as expression, block expressions, and `&&`/`||` operators to reduce overhead when this extension is turned off

**Review by tats-u — CHANGES_REQUESTED** [review]
The current change is still far from official recognition. While I haven't thoroughly understood the pulldown-cmark codebase yet, there's still room for improvement—including areas beyond what I've pointed out—and the test cases are at best inadequate.

**Comment by tats-u** [comment]
pulldown-cmark is one of the fastest Markdown parsers, so we need to keep the performance when this extension is turned off. We may have to introduce CodSpeed to monitor performance regression in advance.

**Comment by 3w36zj6** [comment]
Thanks for the review.

For the tests, I changed them to follow the same pattern as the other extensions: the suite is generated from a spec fixture. The source URLs I used are listed explicitly with commit hashes. While doing that, I also fixed the boundary checks for CJK-friendly emphasis cases involving variation selectors. The test coverage should now be much more comprehensive.

For the performance concern, I updated the implementation so that, when CJK-friendly emphasis is disabled, the extra work for this extension is avoided as much as possible.

File: pulldown-cmark/src/firstpass.rs [hunk 11]
```diff
@@ -2523,31 +2487,105 @@ fn delim_run_can_close(
...
-        is_prev_non_cjk_punctuation_sequence(prev_char_opt, prev_prev_char)
-    } else {
-        is_punctuation(prev_char)
-    };
+    if cjk_friendly_emphasis && matches!(delim, b'*' | b'_') {
+        let mut previous_chars = s[..ix].chars().rev();
+        let _ = previous_chars.next();
+        return cjk_friendly_delim_run_can_close(
+            prev_char,
+            previous_chars.next(),
```
Comment (tats-u) [inline pulldown-cmark/src/firstpass.rs]: Nitpick/perf: this character is actually not used in most cases. I wonder if `LazyCell` is too slow (has too much overhead).

File: pulldown-cmark/src/firstpass.rs [hunk 12]
```diff
@@ -2488,11 +2488,13 @@ fn delim_run_can_close(
...
-        let mut previous_chars = s[..ix].chars().rev();
-        let _ = previous_chars.next();
         return cjk_friendly_delim_run_can_close(
             prev_char,
-            previous_chars.next(),
+            || {
+                let mut previous_chars = s[..ix].chars().rev();
+                let _ = previous_chars.next();
+                previous_chars.next()
+            },
```
Comment (tats-u) [inline pulldown-cmark/src/firstpass.rs]:
```suggestion
            || s[..(ix - prev_char.len_utf8())].chars().rev(),
```

File: pulldown-cmark/src/cjk.rs [hunk 13]
```diff
@@ -0,0 +1,151 @@
+use crate::puncttable::is_punctuation;
+
+/// Returns true when `ch` is considered a CJK character by the
+/// CommonMark CJK-friendly amendments specification.
+///
+/// These ranges are based on Unicode version 17.
+/// The source is https://github.com/tats-u/markdown-cjk-friendly/blob/ee93f3c2dffc8c7eb25c7a1c9f3f962279ac247d/ranges.md.
+pub(crate) fn is_cjk_character(ch: char) -> bool {
+    let cp = ch as u32;
+    matches!(
```
Comment (tats-u) [inline pulldown-cmark/src/cjk.rs:13]: I prefer a fast path:

```suggestion
    if cp < 0x1100 {
        return false;
    }
    matches!(
```

File: pulldown-cmark/src/firstpass.rs [hunk 14]
```diff
@@ -2472,6 +2487,18 @@ fn delim_run_can_close(
         }
     }
     let delim = suffix.bytes().next().unwrap();
+    if cjk_friendly_emphasis && matches!(delim, b'*' | b'_') {
```
Comment (tats-u) [inline pulldown-cmark/src/firstpass.rs]: `~~` should be CJK friendly too.

File: pulldown-cmark/src/firstpass.rs [hunk 15]
```diff
@@ -2417,6 +2428,9 @@ fn delim_run_can_open(
             return false;
         }
     }
+    if cjk_friendly_emphasis && matches!(delim, b'*' | b'_') {
```
Comment (tats-u) [inline pulldown-cmark/src/firstpass.rs]: `~~` should be CJK friendly too.

File: pulldown-cmark/src/firstpass.rs [hunk 16]
```diff
@@ -2486,6 +2513,84 @@ fn delim_run_can_close(
...
+        close
+    } else {
+        close && (after_space_or_punctuation || !open)
+    };
+
+    CjkFriendlyDelimiterRunFlanking {
+        can_open,
+        can_close,
+    }
+}
```
Comment (tats-u) [inline pulldown-cmark/src/firstpass.rs]: I updated the "implementers tips":

- https://github.com/tats-u/markdown-cjk-friendly/blob/main/implementers-tips.md#flowchart-to-suppress-extra-processes
- https://github.com/tats-u/markdown-cjk-friendly/blob/main/implementers-tips.md#logical-expressions-transformations

You don't need to use the results whether characters are "non-CJK punctuation". Instead of that  you can check whether they are CJK characters or punctuation.

File: pulldown-cmark/specs/cjk_friendly_emphasis.txt [hunk 17]
```diff
@@ -0,0 +1,474 @@
...
+~~真，~~她
+
+This~~is~~stricken
+
+Here I strike out an exclamation point~~!~~.
+.
+<p>真<del>（她</del></p>
+<p><del>真，</del>她</p>
+<p>This<del>is</del>stricken</p>
+<p>Here I strike out an exclamation point<del>!</del>.</p>
```
Comment (tats-u) [inline pulldown-cmark/specs/cjk_friendly_emphasis.txt:470]: https://tats-u.github.io/markdown-cjk-friendly/?sc8=55yffn7vvIjlpbl-fgoKfn7nnJ_vvIx-fuWluQoKVGhpc35-aXN-fnN0cmlja2VuCgpIZXJlIEkgc3RyaWtlIG91dCBhbiBleGNsYW1hdGlvbiBwb2ludH5-IX5-Lg&gfm=1&engine=micromark&ver=2.0.1

```suggestion
<p>真<del>（她</del></p>
<p><del>真，</del>她</p>
<p>This<del>is</del>stricken</p>
<p>Here I strike out an exclamation point~~!~~.</p>
```

File: pulldown-cmark/tests/suite/cjk_friendly_emphasis.rs [hunk 18]
```diff
@@ -432,6 +432,25 @@ _abc_漢
...
+~~真，~~她
+
+This~~is~~stricken
+
+Here I strike out an exclamation point~~!~~.
+"##;
+    let expected = r##"<p>真<del>（她</del></p>
+<p><del>真，</del>她</p>
+<p>This<del>is</del>stricken</p>
+<p>Here I strike out an exclamation point<del>!</del>.</p>
```
Comment (tats-u) [inline pulldown-cmark/tests/suite/cjk_friendly_emphasis.rs:458]: https://tats-u.github.io/markdown-cjk-friendly/?sc8=55yffn7vvIjlpbl-fgoKfn7nnJ_vvIx-fuWluQoKVGhpc35-aXN-fnN0cmlja2VuCgpIZXJlIEkgc3RyaWtlIG91dCBhbiBleGNsYW1hdGlvbiBwb2ludH5-IX5-Lg&gfm=1&engine=micromark&ver=2.0.1

```suggestion
<p>Here I strike out an exclamation point~~!~~.</p>
```

**Comment by tats-u** [comment]
`~~` behaves like `**`, but both maintain traditional non-CJK-friendly behavior when wrapping non-CJK punctuation within non-CJK letters.

**Comment by tats-u** [comment]
> This branch has conflicts that must be resolved

File: pulldown-cmark/src/firstpass.rs [hunk 19]
```diff
@@ -2620,7 +2616,7 @@ fn cjk_friendly_delim_run_flanking(
             || after_cjk
             || matches!(prev_char, '*' | '_'));
 
-    let can_open = if delim == b'*' {
+    let can_open = if matches!(delim, b'*' | b'~') {
```
Comment (tats-u) [inline pulldown-cmark/src/firstpass.rs]:
```suggestion
    let can_open = if cjk_changes_flanking {
```

File: pulldown-cmark/src/firstpass.rs [hunk 20]
```diff
@@ -2598,7 +2594,7 @@ fn cjk_friendly_delim_run_flanking(
     }
 
     let prev_sequence = classify_preceding_cjk_friendly_sequence(prev_char, get_prev_prev_char);
-    let cjk_changes_flanking = delim == b'*';
+    let cjk_changes_flanking = matches!(delim, b'*' | b'~');
```
Comment (tats-u) [inline pulldown-cmark/src/firstpass.rs]: @ollpu Do you think there are delimiter marks other than emphasis (`_` / `*`) and strikethrough (`~`) that behave like `_`?

If not, this should be changed to:

```suggestion
    let cjk_changes_flanking = delim != b'_';
```

**Review by tats-u — APPROVED** [review]
The latter part seems adequate as an initial implementation of this extension.

**Comment by tats-u** [comment]
@Martin1887 Do you have time to check this PR? I, the original author of this extension, believe the content is fine.

**Comment by Martin1887** [comment]
Thanks for the reminder. It is a bit quite large, but I will try to review it in the following days. I also add @notriddle as reviewer, who will find more technical details and improvements in case they exist.

**Comment by tats-u** [comment]
Thank you for your reply. I hope playgrounds will help your review:

Official playground: https://tats-u.github.io/markdown-cjk-friendly/ (marked plugin seems to have poorer quality than those for the others)
Comrak (GLFM) playground: https://gitlab-org.gitlab.io/ruby/gems/gitlab-glfm-markdown/

File: pulldown-cmark/src/firstpass.rs [hunk 21]
```diff
@@ -32,6 +36,7 @@ pub(crate) fn run_first_pass(text: &str, options: Options) -> (Tree<Item>, Alloc
         last_line_blank: false,
         allocs: Allocations::new(),
         options,
+        cjk_friendly_emphasis: options.contains(Options::ENABLE_CJK_FRIENDLY_EMPHASIS),
```
Comment (notriddle) [inline pulldown-cmark/src/firstpass.rs]: If a function already accepts `options: Options`, you should not add a `cjk_friendly_emphasis: bool` parameter.

Functions that accept `bool`ean parameters for extensions, like `scan_paragraph_interrupt_no_table`, do not have an `options: Options` parameter.

One or the other. Not both.

**Review by notriddle — CHANGES_REQUESTED** [review]

**Comment by notriddle** [comment]
I found a few strikethrough-related problems, and pushed the fixes to a new commit that you can pull in if they look good.

https://github.com/notriddle/pulldown-cmark/commit/4c0777a30794d566359ced599852d0ca7c68029b

**Comment by tats-u** [comment]
It seems this PR or your suggestion has made some changes to the single tilde syntax. While I completely agree that the single tilde syntax should be less likely to produce strikethrough effects compared to the double tilde syntax, it will also have to be reflected in my original specifications (https://github.com/tats-u/markdown-cjk-friendly/blob/main/specification.md).

**Comment by tats-u** [comment]
From my understanding, your proposal is to change the single tilde syntax to match the conventional CJK-unfriendly asterisk syntax. Is that correct?

**Comment by tats-u** [comment]
I know the combination of CJK-friendly extension and the single tilde syntax causes some problems by https://github.com/vercel/streamdown/issues/545.

**Comment by notriddle** [comment]
The way it’s supposed to work in pulldown-cmark is described here: https://pulldown-cmark.github.io/pulldown-cmark/specs/strikethrough.html

> From my understanding, your proposal is to change the single tilde syntax to match the conventional CJK-unfriendly asterisk syntax. Is that correct?

Single tilde strikeout in pulldown-cmark is supposed to act likes underscores, not asterisks. In CJK-friendly mode, single tilde strikeout is supposed to act like CJK-friendly underscores.

**Comment by tats-u** [comment]
> Single tilde strikeout in pulldown-cmark is supposed to act likes underscores, not asterisks.

Ah, I see.

**Comment by tats-u** [comment]
I don't see any issues with merging https://github.com/notriddle/pulldown-cmark/commit/4c0777a30794d566359ced599852d0ca7c68029b. Go ahead and do it.

File: pulldown-cmark/src/firstpass.rs (hunk 20 above)
Comment (notriddle) [reply pulldown-cmark/src/firstpass.rs]: Is it even possible for this to evaluate to false here?

The `is_cjk_friendly_delim` function only returns true if `delim` is `*`, `_`, or `~`. On line 2544, we early return from this function when `delim` is `_` or `~`. That should mean that `delim` will always be `*` at this point.

Comment (tats-u) [reply pulldown-cmark/src/firstpass.rs]: You're right. `cjk_changes_flanking` is always `true`.

**Review by notriddle — APPROVED** [review]

**Comment by guoliu** [comment]
An adopter data point, and a thank-you 🙏

[moss](https://github.com/Symbiosis-Lab/moss-releases) (a Rust + Tauri static-site generator & editor) hit exactly this on a published Chinese-language article: `**中文。**后面` — and every similar sentence — rendered as literal `**`, because the closing delimiter falls between CJK punctuation and an ideograph with no ASCII space between them.

We validated this PR against our real content and pipeline. With `Options::ENABLE_CJK_FRIENDLY_EMPHASIS`:

- all the affected spans render correctly, and
- our existing CommonMark/ASCII emphasis tests are unchanged — backward-compatible, exactly as the [`tats-u/markdown-cjk-friendly`](https://github.com/tats-u/markdown-cjk-friendly) spec promises.

We're shipping it now via a `[patch.crates-io]` fork of this branch, and would love to see it land and released so we can drop the patch. Happy to help test on real CJK corpora. Thanks @3w36zj6 for implementing the spec, and to the maintainers for reviewing it.

**Comment by Martin1887** [comment]
Please fix the errors in tests detected in CI before merging this pull request. Thanks!
